### DISCUSSION TITLE
`ProductType` and `Attribute` relations

### DISCUSSION THREAD

**Discussion by netzknecht** [lunarphp/lunar#1346] [Q&A]
I think it might be useful to be able to map `Attribute` models to the `ProductType` model, as well, but there are some aspect to be cleared for me.

Currently there are two different types to define the direct relation to `Attribute` models with the `mappedAttributes` method. 

1. [As `HasMany` relation in `HasAttributes` trait](https://github.com/lunarphp/lunar/blob/abc1db6cec689fc4a9433863c795c0646b05640a/packages/core/src/Base/Traits/HasAttributes.php#L23C25-L23C25)
2. [As `MorphToMany` relation in `Brand` model for example ](https://github.com/lunarphp/lunar/blob/abc1db6cec689fc4a9433863c795c0646b05640a/packages/core/src/Models/Brand.php#L61)

Even if both implementations produce the same result, in respect of code design and to prevent redundancy, the use of the `HasAttributes` trait with the as `HasMany` relationship should be preferred to define the direct relationship.

Additionally, attribute casting could be implemented in the trait:

```
    /**
     * Method when trait is initialized.
     *
     * @return void
     */
    public function initializeHasAttributes(): void
    {
        $this->mergeCasts([
            'attribute_data' => AsAttributeData::class,
        ]);
    }
```

There is also a naming conflict regarding the relation of `Attribute` models to the `ProductType` model, because the `mappedAttributes` method is used to define the polymorphic `MorphToMany` relation to be able to restrict the `Attribute` models mapped to `Product` and `ProductVariant` model dependent on the `ProductType`. If we rename this relation to `attributables` for example, it would be possible to use the `HasAttributes` trait to map `Attribute` models to `ProductType` model and the required polymorphic `MorphToMany` relation as well.

**Comment by netzknecht** [comment]
The part "[Attributes: Advanced usage](https://docs.lunarphp.io/core/reference/attributes.html#advanced-usage)" of the docs  therefore does not work as it could be interpreted.

**Comment by netzknecht** [comment]
Check out my [fix/mapped-attributes-and-attributables](https://github.com/netzknecht/lunarphp-lunar/tree/fix/mapped-attributes-and-attributables) branch, it should cover everything.

**Reply by glennjacobs** [reply]
Thanks, we'll take a look once we come up for air from the Filament development :-)

**Reply by glennjacobs** [reply]
Perhaps do a draft PR, so it doesn't fall off our radar?

**Reply by netzknecht** [reply]
> Thanks, we'll take a look once we come up for air from the Filament development :-)

Therefore, in the hub package, I made only absolutely necessary adjustments and largely retained the naming conventions there.

**Reply by netzknecht** [reply]
> Perhaps do a draft PR, so it doesn't fall off our radar?

Done #1347
